src/riverford/durable_ref/scheme/redis/carmine.clj

```diff
@@ -0,0 +1,152 @@
+(ns riverford.durable-ref.scheme.redis.carmine
+  "Carmine-backed Redis implementation.
+
+  Supports the following reference types:
+
+  - atomic
+  - value
+  - volatile
+
+  References follow the scheme <reftype>:redis:tcp://<host>:<port>/<dbnumber>/<key>.<format>, e.g.:
+
+  atomic:redis:tcp://localhost:6379/0/atom-ref.edn
+
+  Keys can be namespaced further, e.g.
+
+  atomic:redis:tcp://localhost:6379/0/atoms/a1-ref.edn
+
+  This results in the top-level atoms key in Redis being a hash map with a single
+  entry a1-ref.edn. A second ref,
+
+  atomic:redis:tcp://localhost:6379/0/atoms/a2-ref.edn
+
+  inserts another entry in the top-level atoms key in Redis, and so on.
+
+  To be able to use Clojure's reference type interfaces you can add the Redis credentials
+  using `add-credentials!`, and remove them again when no longer needed with `remove-credentials!`.
+  Example, with Redis password \"foobar\":
+
+  `(add-credentials! \"localhost\" 6379 \"foobar\")`
+  `(def ref-1 (dref/reference \"atomic:redis:tcp://localhost:6379/0/ref-1.edn\"))`
+  `@ref-1 ;=> nil`
+  `(reset! ref-1 {:foo #{:a :b}})`
+  `@ref-1 ;=> {:foo #{:a :b}}`
+  `(swap! ref-1 assoc :bar 42)`
+  `@ref-1 ;=> {:foo #{:a :b} :bar 42}`
+  "
+  (:require
+    [clojure.edn :as edn]
+    [clojure.string :as str]
+    [riverford.durable-ref.core :as dref]
+    [taoensso.carmine :as car])
+  (:import
+    (java.net URI)))
+
+(defonce connections (atom {}))
+(defonce credentials (atom {}))
+
+(defn add-credentials!
+  [host port password]
+  (swap! credentials assoc (str host ":" port) {:password password}))
+
+(defn remove-credentials!
+  [host port]
+  (swap! credentials dissoc (str host ":" port)))
+
+(defn- uri->connection
+  [^URI uri]
+  (let [sub-uri (URI. (.getSchemeSpecificPart uri))
+        path (.getPath sub-uri)]
+    (if-let [connection (get @connections (.getHost sub-uri))]
+      connection
+      (let [new-connection {:pool {}
+                            :spec {:host (.getHost sub-uri)
+                                   :port (let [port (.getPort sub-uri)]
+                                           (if (pos? port)
+                                             port
+                                             6379))}}]
+        (swap! connections assoc (str (get-in new-connection [:spec :host]) ":" (get-in new-connection [:spec :port])) new-connection)
+        new-connection))))
+
+(defn- location
+  [^URI uri]
+  (let [sub-uri (URI. (.getSchemeSpecificPart uri))
+        path (.getPath sub-uri)
+        [_ db id-or-ns ?id] (str/split path #"/")]
+    [(edn/read-string db)
+     (when ?id id-or-ns)
+     (or ?id id-or-ns)]))
+
+(defn- get-credentials
+  [opts connection]
+  (let [c1 (:credentials opts)
```

Comment:
I think I would prefer `(-> opts :scheme :redis :carmine :credentials)` to be consistent with the Dynamo/S3 scheme option passing style.

Reply:
Of course, I actually intended to follow that style, but that must have slipped.